lib/src/results.dart

```diff
@@ -207,3 +210,33 @@ extension RealmResultsInternal on RealmResults {
     return RealmResults<T>._(handle, realm);
   }
 }
+
+class _RealmResultsIterator<T extends RealmObject> implements Iterator<T> {
+  final RealmResults<T> _results;
+  final int _length;
+  int _index;
+  T? _current;
+
+  _RealmResultsIterator(RealmResults<T> results)
+      : _results = results,
+        _length = results.length,
+        _index = 0;
+
+  @override
+  T get current => _current as T;
+
+  @override
+  bool moveNext() {
+    int length = _results.length;
+    if (_length != length) {
```

Comment:
I would consider dropping `_length field` and just check against `_index`, ie. `_index >= length`. This would allow more modifications to the result set, even if the produced sequence will be odd. The current check doesn't catch all modifications anyway, but maybe check some of the other SDKs?

Reply:
Ie. somthing like:
```
  @override
  bool moveNext() {
    int length = _results.length;
    if (_index == length) return false;
    if (_index > length) {
      throw ConcurrentModificationError(_results);
    }
    _current = _results[_index];
    _index++;
    return true;
  }

```

Reply:
I think it is better to have some notification when something has changed in the result concurrently while it is iterated. Otherwise we could break some of the logic in the apps. Let's discuss it and decide how to do it.

Reply:
Could even consider just returning `false` on `_index >= length`?

Reply:
We won't see all changes anyway. If the i'th index in the result set is changed, the current check won't catch that. I think I prefer to be as robust as possible here. Sync may change a realm result, and I don't think users will like to guard all iterations against `ConcurrentModificationError`.. but I agree - it is not optimal

Reply:
Follow up on zoom discussion.. to match .net would be:@override
```
  bool moveNext() {
    int length = _results.length;
    if (_index >= length) {
      _current = null;
      return false;
    }
    _current = _results[_index];
    _index++;
    return true;
  }
```
.. but also snapshot (if possible) the collection before iteration

Reply:
For some background on the .NET code - while this is a departure from the BCL, we felt it's warranted because of the nature of Realm. Getting `ConcurrentModificationError` due to sync/some background thread modifying the collection while you're iterating over it is going to be super unintuitive for users and as Kasper mentions, most users will not think of adding error handling to each foreach (which is understandable considering how annoying it will be), meaning throwing here will likely result in a hard crash in production. So my recommendation would be not to throw and add snapshotting if possible.

Reply:
I think we should go with the above, and look into snapshotting later

Reply:
Btw, in .Net you will also receive exception at runtime if you are trying to change collection while iterate. I think it is expected. Otherwise we may introduce unpredictable and hard to detect problems. Let me think a little bit.

Reply:
Yes for the BCL, but not for `RealmCollectionBase`. As @nirinchev state, that is a deliberate choice.

Reply:
Yes, it is acceptable not to throw exceptions but only in case we have a snapshot, otherwise would be a problem.

Reply:
@nirinchev and @nielsenko thank you for the comments! I removed throwing concurrent exception. We will make a snapshot with issue #166. May I resolve this conversation?